sycl/doc/extensions/proposed/sycl_ext_oneapi_non_uniform_groups.asciidoc

```diff
@@ -0,0 +1,1161 @@
+or may be in a preliminary state.  The specification itself may also change in
+incompatible ways before it is finalized.  *Shipping software products should
+not rely on APIs defined in this specification.*
+
+
+== Overview
+
+Many modern hardware architectures support flexible sub-divisions of
+work-groups and sub-groups to support fine-grained work scheduling. A common
+use-case for such flexibility is communication and synchronization between
+work-items in divergent control flow.
+
+This proposal introduces new classes to represent sub-divisions of SYCL's
+built-in group types, traits for detecting these classes, and free functions
+for creating new instances of these classes.
+
+These new classes can be used as arguments to group functions, group algorithms,
+and custom functions to convey exactly which work-items an operation is expecting,
+simplifying interfaces and greatly reducing the amount of delicate wording
+needed in function documentation.
+
+NOTE: The first version of this extension only supports partitioning of
+sub-groups. It is expected that in the future, these functions will be expanded
+to also allow partitioning of root-groups, work-groups and user-constructed
+groups.
+
+
+== Specification
+
+=== Feature test macro
+
+This extension provides a feature-test macro as described in the core SYCL
+specification.  An implementation supporting this extension must predefine the
+macro `SYCL_EXT_ONEAPI_NON_UNIFORM_GROUPS` to one of the values defined in the
+table below.  Applications can test for the existence of this macro to
+determine if the implementation supports this feature, or applications can test
+the macro's value to determine which of the extension's features the
+implementation supports.
+
+[%header,cols="1,5"]
+|===
+|Value
+|Description
+
+|1
+|The APIs of this experimental extension are not versioned, so the
+ feature-test macro always has this value.
+|===
+
+=== Control Flow
+
+The SYCL specification defines
+link:https://registry.khronos.org/SYCL/specs/sycl-2020/html/sycl-2020.html#control-flow[control
+flow] as below:
+
+> When all work-items in a group are executing the same sequence of statements,
+> they are said to be executing under _converged_ control flow. Control flow
+> _diverges_ when different work-items in a group execute a different sequence of
+> statements, typically as a result of evaluating conditions differently (e.g.
+> in selection statements or loops).
+
+This extension introduces some new terminology to describe other kinds of
+control flow, to simplify the description of the behavior for new group types.
+
+A work-item is said to have _active_ control flow for a given point of
+execution if it follows the sequence of statements currently being executed.
+
+Two or more work-items are considered to have _simultaneous_ execution at a given
+point if they execute together. The amount of and granularity of simultaneous
```

Comment:
I want to avoid "concurrently" because of the forward progress guarantee with the same name.  You're right that "together" is vague.  I think there may be implementations that execute "together" but aren't SIMD, but I agree that adding it as an example would help.  I'll make this change.